relay-metrics/src/aggregator/inner.rs

```diff
@@ -0,0 +1,926 @@
+                // Break down the maximum project cost to a maximum cost per partition.
+                max_partition_project: config
+                    .max_project_key_bucket_bytes
+                    .map(|c| c.div_ceil(total_slots as u64))
+                    .unwrap_or(u64::MAX),
+            },
+            slot_diff,
+            partition_by: FlushBatching::Partition,
+            hasher: build_hasher(),
+        }
+    }
+
+    /// Returns the configured bucket interval.
+    pub fn bucket_interval(&self) -> u32 {
+        self.bucket_interval
+    }
+
+    /// Returns total aggregator stats.
+    pub fn stats(&self) -> Stats {
+        Stats {
+            count: self.stats.count,
+            count_by_namespace: self.stats.count_by_namespace,
+            cost: self.stats.count,
+            cost_by_namespace: self.stats.cost_by_namespace,
+        }
+    }
+
+    /// Returns `true` if the aggregator contains any metric buckets.
+    pub fn is_empty(&self) -> bool {
+        self.stats.cost == 0
+    }
+
+    /// Returns the time as a duration since epoch when the next flush is supposed to happen.
+    pub fn next_flush(&self) -> Duration {
+        // `head + 1` to get the end time of the slot not the start.
+        (Duration::from_secs(self.head + 1) / self.num_partitions * self.bucket_interval)
+            + Duration::from_secs(u64::from(self.delay))
+    }
+
+    /// Merges a metric bucket.
+    ///
+    /// Returns `true` if the bucket already existed in the aggregator
+    /// and `false` if the bucket did not exist yet.
+    pub fn merge(
+        &mut self,
+        mut key: BucketKey,
+        value: BucketData,
+    ) -> Result<(), AggregateMetricsError> {
+        let project_key = key.project_key;
+        let namespace = key.metric_name.namespace();
+
+        let time_slot = key.timestamp.as_secs() / u64::from(self.bucket_interval);
+        // Make sure the timestamp is normalized to the correct interval as well.
+        key.timestamp = UnixTimestamp::from_secs(time_slot * u64::from(self.bucket_interval));
+
+        let now_slot = self.head / u64::from(self.num_partitions);
+        if !self.slot_diff.contains(now_slot, time_slot) {
+            return Err(AggregateMetricsError::InvalidTimestamp(key.timestamp));
+        }
+
+        let assigned_partition = match self.partition_by {
+            FlushBatching::None => 0,
+            FlushBatching::Project => self.hasher.hash_one(key.project_key),
+            FlushBatching::Bucket => self.hasher.hash_one(&key),
+            FlushBatching::Partition => {
+                self.hasher
+                    .hash_one((key.project_key, &key.metric_name, &key.tags))
+            }
+        } % u64::from(self.num_partitions);
+
+        let slot = time_slot * u64::from(self.num_partitions) + assigned_partition;
+
+        let slots_len = self.slots.len() as u64;
+        let index = (slot + slots_len).wrapping_sub(self.head % slots_len) % slots_len;
```

Comment:
What happens with different time stamps that map to the same index? Is that case prevented by how `slots.len()` is chosen in the beginning?

Reply:
This is fine and will happen for backdated und future buckets, there are tests that cover this case. Since the timestamp is part of the key, they won't be aggregated together and are independent.